### PR TITLE
feat: v1.1.0 — LLM skill, compact responses, move_file, SEA binary

### PR DESCRIPTION
**Pull request by andreinknv**
## Summary

- **LLM Skill resource** (`obsidian://skill`): Battle-tested MCP resource with 7 sections — golden rules, step-by-step workflows, error recovery, tool selection guide, known pitfalls, consolidated action reference (mode-dependent), compact field mapping (config-dependent). Also ships as `.claude/skills/obsidian-mcp/SKILL.md` in the npm package.
- **Compact responses**: `OBSIDIAN_COMPACT_RESPONSES=true` maps verbose field names to abbreviations (`content`→`c`, `frontmatter`→`fm`, `path`→`p`, etc.) and strips JSON whitespace. Runtime-toggleable via `configure` tool. Env var: 17→18 total.
- **move_file tool**: Granular tool #9 / consolidated `vault` action `move`. Copies content to destination, deletes source to Obsidian trash. Conflict detection, no-op for same-path. In `full` (39 tools) and `safe` (35 tools) presets.
- **SEA binary**: `npm run build:sea` compiles a standalone binary via Node.js Single Executable Applications (esbuild bundle + postject inject + macOS codesign).

## Changes

| File | Change |
|------|--------|
| `src/skill.ts` | **New** — `buildSkillContent(mode, compact)` with 7 sections |
| `.claude/skills/obsidian-mcp/SKILL.md` | **New** — Static skill file (sections 1-5, granular default) |
| `src/config.ts` | Add `compactResponses` to Config, DEFAULTS, schema, loader, redacted |
| `src/obsidian.ts` | Add `compactify()`, `setCompactResponses()`, update `jsonResult()` |
| `src/index.ts` | Register MCP resource, call `setCompactResponses()` |
| `src/tools/shared.ts` | Add `handleMoveFile()`, configure tool handles `compactResponses` |
| `src/tools/granular.ts` | Register `move_file` tool (#9) |
| `src/tools/consolidated.ts` | Add `move` action to `vault` tool |
| `src/tools.ts` | Add `move_file` to `full` and `safe` presets |
| `scripts/build-sea.js` | **New** — SEA build script |
| `package.json` | Version 1.1.0, `build:sea` script, `files` includes `.claude/skills`, esbuild/postject devDeps |
| `CHANGELOG.md` | v1.1.0 entry |
| Tests | 636 tests passing, 89%+ coverage, new tests for all features |

## Test plan

- [x] `npm run build` — zero errors
- [x] `npm run lint` — zero errors
- [x] `npm run test` — 636 tests passing
- [x] `npm run test:coverage` — 89.2% overall (threshold: 80%)
- [x] `npm audit` — zero vulnerabilities
- [x] `npx madge --circular --extensions ts src/` — zero circular deps
- [x] CodeRabbitAI review — 0 actionable, 0 nitpicks, all 16 threads resolved
- [x] Greptile review — 5/5 confidence, all Fix-All items addressed
- [x] Sonar — 0 bugs, 0 smells, 0 vulnerabilities, 88.3% coverage
- [x] Manual: test skill resource for both modes (granular + consolidated, compact on/off) — 32/32 passed
- [x] Manual: test compact responses on/off (field mapping, opaque keys, whitespace) — verified live
- [x] Manual: test move_file edge cases (normal, no-op, conflict, 404, non-.md, subdirectory, path traversal, cache) — 25/25 passed
- [x] Manual: SKILL.md golden rules + workflows against live Obsidian — 29/30 passed (1 false-negative)
- [x] Manual: cache graph analysis (backlinks, connections, orphans, structure, invalidation) — all passed
- [x] `npm run test:smoke` — 9/9 passed against live mcp-test-vault
- [x] Stress tests — 10/10 passed, 225 ops in 700ms
- [x] `npm publish --dry-run` — 48 files, 101.3 kB, no secrets

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * File move/rename support for vault operations with conflict and error handling
  * Compact-response mode (config/env toggle) to produce smaller, field-mapped JSON outputs
  * Obsidian MCP skill usage guide exposed via a server resource (markdown)
  * Standalone executable build and release workflow

* **Documentation**
  * Changelog and spec updated to document new features and modes
  * Added comprehensive skill usage guide

* **Tests**
  * Expanded tests for compact responses, skill content, and move-file behavior

* **Chores**
  * Version bumped to 1.1.0; added build tooling and scripts

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -268,6 +270,14 @@ async function main(): Promise<void> {
 
   const toolCount = registerAllTools(server, client, cache, config);
 
+  const skillContent = buildSkillContent(config.toolMode, config.compactResponses);
+  server.registerResource(
+    "obsidian-skill",
+    "obsidian://skill",
+    { description: "LLM usage guide for Obsidian MCP tools" },
+    async () => ({ contents: [{ uri: "obsidian://skill", text: skillContent, mimeType: "text/markdown" }] }),
+  );
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/index.ts]: Fixed in de1c686. Skill resource callback now builds content on each read using live getCompactResponses() state.

File: src/tools/shared.ts [hunk 2]
```diff
@@ -407,6 +409,41 @@ export async function batchGetFiles(
...
+ * @returns A tool result describing success, no-op, or conflict.
+ */
+export async function handleMoveFile(
+  client: ObsidianClient,
+  source: string,
+  destination: string,
+): Promise<ToolResult> {
+  if (source === destination) {
+    return textResult(`No-op: source and destination are the same (${source})`);
+  }
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/tools/shared.ts]: Fixed in de1c686. handleMoveFile now calls sanitizeFilePath() on both paths before comparison.

File: src/obsidian.ts [hunk 3]
```diff
@@ -243,6 +243,71 @@ function encodeTargetHeader(target: string): string {
...
+]);
+
+/**
+ * Recursively maps known field names to compact abbreviations and strips null/undefined values.
+ * @param data - The data to compactify.
+ * @returns A new object with compact field names.
+ */
+export function compactify(data: unknown): unknown {
+  if (data === null || data === undefined) return undefined;
+  if (Array.isArray(data)) return data.map(compactify).filter((v) => v !== undefined);
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/obsidian.ts]: Fixed in de1c686. compactify now preserves null, only strips undefined. Added OPAQUE_KEYS (frontmatter, result) — their contents are not recursed into.

File: scripts/build-sea.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  };
+  writeFileSync(CONFIG_FILE, JSON.stringify(seaConfig));
+
+  // 3. Generate blob
+  run(`node --experimental-sea-config ${CONFIG_FILE}`);
+
+  // 4. Copy node binary
+  // eslint-disable-next-line no-console -- build script, not MCP transport
+  console.log("Step 3: Creating binary...");
+  copyFileSync(process.execPath, BINARY_NAME);
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply scripts/build-sea.js]: Fixed in de1c686. Now uses process.execPath for blob generation.

File: src/__tests__/obsidian.test.ts [hunk 5]
```diff
@@ -121,6 +124,99 @@ describe("jsonResult", () => {
...
+// setCompactResponses / getCompactResponses + jsonResult compact mode
+// ---------------------------------------------------------------------------
+describe("compact responses", () => {
+  afterEach(() => {
+    setCompactResponses(false);
+  });
+
+  it("defaults to disabled", () => {
+    expect(getCompactResponses()).toBe(false);
+  });
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/__tests__/obsidian.test.ts]: Fixed in de1c686. Tests now save/restore compact state in beforeEach/afterEach.

File: src/config.ts [hunk 6]
```diff
@@ -331,6 +336,7 @@ export function getRedactedConfig(config: Config): Record<string, unknown> {
     excludeTools: config.excludeTools,
     cacheTtl: config.cacheTtl,
     enableCache: config.enableCache,
+    compactResponses: config.compactResponses,
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/config.ts]: Fixed in de1c686. getRedactedConfig accepts overrides param; handleConfigureShow passes live getCompactResponses().

File: src/index.ts [hunk 7]
```diff
@@ -268,6 +270,14 @@ async function main(): Promise<void> {
 
   const toolCount = registerAllTools(server, client, cache, config);
 
+  const skillContent = buildSkillContent(config.toolMode, config.compactResponses);
+  server.resource(
+    "obsidian-skill",
+    "obsidian://skill",
+    { description: "LLM usage guide for Obsidian MCP tools" },
+    async () => ({ contents: [{ uri: "obsidian://skill", text: skillContent, mimeType: "text/markdown" }] }),
+  );
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/index.ts]: Fixed in de1c686. Skill content is now built on each resource read, not captured at startup.

File: src/obsidian.ts [hunk 8]
```diff
@@ -243,6 +243,71 @@ function encodeTargetHeader(target: string): string {
...
+  if (typeof data === "object") {
+    const obj = data as Record<string, unknown>;
+    const result: Record<string, unknown> = {};
+    for (const key of Object.keys(obj)) {
+      const val = compactify(obj[key]);
+      if (val === undefined) continue;
+      const mappedKey = COMPACT_FIELD_MAP.get(key) ?? key;
+      result[mappedKey] = val;
+    }
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/obsidian.ts:313]: Fixed in de1c686. Added OPAQUE_KEYS set (frontmatter, result) — passed through without recursion. Compact mapping only applies to protocol envelope fields.

File: src/skill.ts [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+  return `## Golden Rules
+
+- ALWAYS get_file_contents(path, format: "map") BEFORE any patch_content — verify the heading exists first. Never patch a heading you haven't confirmed.
+- ALWAYS get_file_contents(path, format: "json") BEFORE modifying frontmatter — see the current state.
+- Use search_replace for precise text changes — safer than put_content which overwrites the entire file.
+- Use batch_get_file_contents for multiple files — never sequential get_file_contents calls.
+- Use get_vault_structure at the start of a session to understand the vault layout (note count, links, orphans, most connected).
+- NEVER use put_content to edit a section — it replaces the ENTIRE file. Use append_content, patch_content, or search_replace instead.
+- NEVER retry a non-idempotent tool on timeout: append_content, patch_content, search_replace, append_active_file, patch_active_file, all append/patch periodic note tools.
+- NEVER assume a path exists — verify with list_files_in_dir or simple_search first.`;
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/skill.ts]: Fixed in de1c686. move_file added to no-retry list in both src/skill.ts and .claude/skills/obsidian-mcp/SKILL.md.

File: src/skill.ts [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+
+  if (mode === "consolidated") {
+    sections.push(consolidatedActionReference());
+  }
+
+  if (compact) {
+    sections.push(compactFieldReference());
+  }
+
+  return `${sections.join("\n\n")}\n`;
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/skill.ts:333]: Intentional: Sections 1-5 use granular tool names because they're more descriptive. In consolidated mode, Section 6 provides the action mapping. Granular names in core sections communicate intent clearly in both modes.

File: src/tools/consolidated.ts [hunk 11]
```diff
@@ -105,6 +108,11 @@ async function handleVaultAction(
   args: VaultArgs,
 ): Promise<ToolResult> {
   if (action === "list") return jsonResult(await client.listFilesInVault());
+  if (action === "move") {
+    if (!args.source) return errorResult("[vault] source is required for move");
+    if (!args.destination) return errorResult("[vault] destination is required for move");
+    return handleMoveFile(client, args.source, args.destination);
+  }
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/tools/consolidated.ts:115]: Fixed in de1c686. Vault tool catch now uses args.source ?? path for move action error context.

File: src/tools/shared.ts [hunk 12]
```diff
@@ -407,6 +417,41 @@ export async function batchGetFiles(
...
+  try {
+    await client.getFileContents(destination, "markdown");
+    return errorResult("CONFLICT: Destination already exists. Delete it first or choose a different path.");
+  } catch (err: unknown) {
+    if (!(err instanceof ObsidianApiError && err.statusCode === 404)) {
+      throw err;
+    }
+  }
+  await client.putContent(destination, content);
+  await client.deleteFile(source);
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/tools/shared.ts]: Fixed in de1c686. deleteFile wrapped in try/catch — returns PARTIAL MOVE error naming both paths on failure.

File: src/tools/shared.ts [hunk 13]
```diff
@@ -230,7 +232,7 @@ function handleConfigureReset(setting: string | undefined, config: Config): Tool
   }
   saveConfigToFile(configPath, resetUpdates);
   // Apply immediately for settings that take effect without restart
-  if (setting === "debug") {
+  if (setting === "debug" || setting === "compactResponses") {
     applyImmediateSetting(setting, "false");
     return textResult(`Setting "${setting}" reset to default (effective immediately)`);
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/tools/shared.ts:237]: Fixed in b7f680c. Reset now uses String(DEFAULTS[setting]) instead of hardcoded "false".

File: src/tools/shared.ts [hunk 14]
```diff
@@ -407,6 +409,41 @@ export async function batchGetFiles(
...
+ * @param client - The Obsidian API client.
+ * @param source - Source file path.
+ * @param destination - Destination file path.
+ * @returns A tool result describing success, no-op, or conflict.
+ */
+export async function handleMoveFile(
+  client: ObsidianClient,
+  source: string,
+  destination: string,
+): Promise<ToolResult> {
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/tools/shared.ts:423]: Already fixed in de1c686. handleMoveFile calls sanitizeFilePath() on both source and destination before any operations.

File: src/skill.ts [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+**404 NOT FOUND** — File doesn't exist.
+- Try adding .md extension if not present
+- Use list_files_in_dir to find the correct path
+- Server has case-insensitive fallback — but path must be close
+
+**PATCH timeout** — Heading/block target not found.
+- Get the document map first: get_file_contents(path, format: "map")
+- Check heading uses :: delimiter for nested headings: "Parent::Child"
+- If heading has special characters, use search_replace instead
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/skill.ts]: Fixed in 21c9928. Changed to 'PATCH 400 error — Heading/block target not found (returned as a bad request, not a timeout).' in both skill.ts and SKILL.md.

File: src/tools/shared.ts [hunk 16]
```diff
@@ -407,6 +407,48 @@ export async function batchGetFiles(
...
+    if (!(err instanceof ObsidianApiError && err.statusCode === 404)) {
+      throw err;
+    }
+  }
+  await client.putContent(normalizedDest, content);
+  try {
+    await client.deleteFile(normalizedSource);
+  } catch (deleteErr: unknown) {
+    const msg = deleteErr instanceof Error ? deleteErr.message : String(deleteErr);
+    return errorResult(`PARTIAL MOVE: File copied to ${normalizedDest} but source ${normalizedSource} could not be deleted (${msg}). Check both paths before retrying.`);
```
[the remark this answers is not in the document]
Comment (andreinknv) [reply src/tools/shared.ts]: Documented in 21c9928. Each client method (getFileContents, putContent, deleteFile) already uses withFileLock internally, serializing per-path writes. MCP tool calls are sequential (single client), so cross-call races are not possible — only one tool handler runs at a time. Added lock ordering comment and rationale. The conflict check remains best-effort since Obsidian itself is the final arbiter.
